### ISSUE TITLE
OPAT category changes when you move from OPAT list view to another list or detail view 

### ISSUE THREAD

**Issue by 7575colli** [openhealthcare/elcid#667]
Not sure how this happened - but i have some pateints stuck in the referrals section and i cant add to all the fields when i open the detailed view. There is also no option to add 'who gives' IV antibiotics.

Help please!

**Comment by davidmiller** [comment]
@7575colli 

This is unfortunate - it didn't come up at all in our testing.

Can you try to reproduce this on http://elcid04-test.herokuapp.com/ ? 

What we need is the steps to reproduce the symptoms - otherwise we can't fix it!

I've converted those two patients to be OPAT, but obviously can't continue to do that for every patient.

@michaeledwardmarks @GabPoll 

Interesting one - Both of these patients were created as OPAT patients, then subsequently changed.
I don't think the user should be able to change the episode type, so some bug has crept in here, caused by behaviour that none of us tested !

**Comment by 7575colli** [comment]
I will try.  Its really odd - im pretty sure that at least one of them was because the pt was 'imported' to OPAT from an inpatient list.  I have added new pt (HM) to referrals this morning - His was imported from an inpt episode or maybe micro list.  This pt is now on OPAT referrals as inpatient too - I cant add 'who gives' antibiotics, also, when i click 'patient details' i cant see any clinical notes, which is a problem for the other pts also noted as inpatients because i can only see their last entry on the main screen and nothing beforehand. Also, to add the pt to the current list, there is no 'accept/refuse' tab on the left - only discharge - which when it is clicked allows the pt to be moved to the current list, i only figured this by experimenting 
There are two other pts on current list noted as inpatients:
LS and AF - can these be changed to OPAT pts too please?

**Comment by 7575colli** [comment]
More strange behaviour - I added pt KK to referrals - the pt was correctly labelled as 'OPAT' and was not imported from a previous episode. 
I then added a diagnosis and PMH - then tried to enter IV antimicrobials, but noticed the patient was now labelled as an 'inpatient' and i cant add 'who gives' and the other mentioned associated problems which is going to make a real mess when it comes to trying to audit

**Comment by michaeledwardmarks** [comment]
I'm testing on Heroku not live system due to internet speed here but I have tried a few things:
Import patient
Add new patient
Save stuff
Can't directly replicate this but see #668 
Would be good if @GabPoll and @davidmiller could investigate with better internet on the UCLH server version

@davidmiller is Heroku-test an exact replica of the current UCLH deployment?

**Comment by 7575colli** [comment]
I cant access Heroku app on the computer that im using.
I just added another pt where the same thing happened.
The sequence seem to be as follows:
-	Pt added to referrals – Initially labelled as OPAT pt in referrals list
-	I added a diagnosis and all was fine
-	I made an enter to PMH which initially wouldn’t save, so I refreshed the page then the pt label had changed to ‘inpatient’

**Comment by michaeledwardmarks** [comment]
Replicated on elCID-LIVE.
Steps as follows:
1) Create patient on OPAT Referral
2) Category == OPAT (CORRECT)
3) Can Add & Save stuff without change to category (CORRECT)
(NB - Doesn't matter if this is a new-to-system patient, or an existing patient on a different list)
4) Go to single patient view 
5) Category shown in Location field on single patient view as "inpatient" (NB lower case "i", INCORRECT)
6) Return to list view
7) Category here also now displaying as "inpatient" (INCORRECT)

Same behaviour can also be triggered by replacing step 4 with switch to any other list.

In general the behaviour seems to be that moving from OPAT-Referral to any other view (different list, or patient details view) changes category to inpatient

Doesn't seem to matter who you view on Patient details e.g if you view the patient you added, or a different patient from the OPAT referrals list. Both result in this bug.

**Comment by michaeledwardmarks** [comment]
Note - same steps do not replicate this on Heroku. 
Therefore this is something that is correct on Heroku but wrong on elCID-Live.

On elCID-Live I didn't seem to see any debugging errors for what it is worth.

**Comment by michaeledwardmarks** [comment]
If you accept a patient to OPAT current then category also becomes inpatient
Difficult to know if that is triggered by the Accept action itself or the fact that, as above, you have moved to a different list view which we now know triggers this behaviour

**Comment by michaeledwardmarks** [comment]
@davidmiller
I think we need to prioritise #667 #668 #669 #670 for a hot-fix - these represent significant bugs interfering with clinical use of system by OPAT on an everyday basis. (with #665 I think = #668)

**Comment by davidmiller** [comment]
This is a race condition that occurs when the category transaction doesn't complete before the read transaction to set tags. Hence the "Works some places not others" (Server configs).

**Comment by davidmiller** [comment]
A fix to these issues has been deployed to elcido4-test - please do take a run at it !

**Comment by michaeledwardmarks** [comment]
@davidmiller

I could never replicate this on Heroku only on elCID Live....not sure how I can test?

**Comment by davidmiller** [comment]
You probably have to trust me and then check it on the live server when we deploy!

I've reproduced it locally, and have a strong mental model for *why* it occurred. 

The fix is here: https://github.com/openhealthcare/opal-opat/commit/b68108ba7292214cb5cd027db9ea768dafdc65e4

Basically, we try to do things in parallel to be faster, and under some circumstances that means that we lose the information about this being an OPAT episode.

We've changed this so we wait for full, final confirmation that we've saved the OPATness before doing everything _else_ in parallel.

It takes a little longer, but now we definitely won't lose the fact that this is an OPAT episode.
